python/modules/IcePy/ObjectAdapter.cpp

```diff
@@ -29,18 +31,18 @@ static unsigned long _mainThreadId;
 namespace IcePy
 {
 
-typedef InvokeThread<Ice::ObjectAdapter> AdapterInvokeThread;
-typedef IceUtil::Handle<AdapterInvokeThread> AdapterInvokeThreadPtr;
-
 struct ObjectAdapterObject
 {
     PyObject_HEAD
     Ice::ObjectAdapterPtr* adapter;
-    IceUtil::Monitor<IceUtil::Mutex>* deactivateMonitor;
-    AdapterInvokeThreadPtr* deactivateThread;
+
+    std::future<void>* deactivateFuture;
+    std::exception_ptr* deactivateException;
     bool deactivated;
-    IceUtil::Monitor<IceUtil::Mutex>* holdMonitor;
-    AdapterInvokeThreadPtr* holdThread;
+
+    std::mutex* holdMutex;
```

Comment:
It would be nice to add some comments.

If I follow the code, we need this holdMutex to "protect" holdXxx, because they can be assigned and reassigned concurrently in different threads. And we can't do that for deactivate.

I think we should simplify in a follow-up PR and eliminate this ability to hold->activate->hold->activate. Why would you ever do that?